cpp/velox/shuffle/VeloxSortBasedShuffleWriter.h

```diff
@@ -0,0 +1,155 @@
+
+#include "VeloxShuffleWriter.h"
+#include "memory/VeloxMemoryManager.h"
+#include "shuffle/PartitionWriter.h"
+#include "shuffle/Partitioner.h"
+#include "shuffle/Utils.h"
+
+#include "utils/Print.h"
+
+namespace gluten {
+
+// set 1 to open print
+#define VELOX_SHUFFLE_WRITER_PRINT 0
+
+#if VELOX_SHUFFLE_WRITER_PRINT
+
+#define VsPrint Print
+#define VsPrintLF PrintLF
+#define VsPrintSplit PrintSplit
+#define VsPrintSplitLF PrintSplitLF
+#define VsPrintVectorRange PrintVectorRange
+#define VS_PRINT PRINT
+#define VS_PRINTLF PRINTLF
+#define VS_PRINT_FUNCTION_NAME PRINT_FUNCTION_NAME
+#define VS_PRINT_FUNCTION_SPLIT_LINE PRINT_FUNCTION_SPLIT_LINE
+#define VS_PRINT_CONTAINER PRINT_CONTAINER
+#define VS_PRINT_CONTAINER_TO_STRING PRINT_CONTAINER_TO_STRING
+#define VS_PRINT_CONTAINER_2_STRING PRINT_CONTAINER_2_STRING
+#define VS_PRINT_VECTOR_TO_STRING PRINT_VECTOR_TO_STRING
+#define VS_PRINT_VECTOR_2_STRING PRINT_VECTOR_2_STRING
+#define VS_PRINT_VECTOR_MAPPING PRINT_VECTOR_MAPPING
+
+#else // VELOX_SHUFFLE_WRITER_PRINT
+
+#define VsPrint(...) // NOLINT
+#define VsPrintLF(...) // NOLINT
+#define VsPrintSplit(...) // NOLINT
+#define VsPrintSplitLF(...) // NOLINT
+#define VsPrintVectorRange(...) // NOLINT
+#define VS_PRINT(a)
+#define VS_PRINTLF(a)
+#define VS_PRINT_FUNCTION_NAME()
+#define VS_PRINT_FUNCTION_SPLIT_LINE()
+#define VS_PRINT_CONTAINER(c)
+#define VS_PRINT_CONTAINER_TO_STRING(c)
+#define VS_PRINT_CONTAINER_2_STRING(c)
+#define VS_PRINT_VECTOR_TO_STRING(v)
+#define VS_PRINT_VECTOR_2_STRING(v)
+#define VS_PRINT_VECTOR_MAPPING(v)
+
+#endif // end of VELOX_SHUFFLE_WRITER_PRINT
+
+class VeloxSortBasedShuffleWriter : public VeloxShuffleWriter {
+ public:
+  static arrow::Result<std::shared_ptr<VeloxShuffleWriter>> create(
+      uint32_t numPartitions,
+      std::unique_ptr<PartitionWriter> partitionWriter,
+      ShuffleWriterOptions options,
+      std::shared_ptr<facebook::velox::memory::MemoryPool> veloxPool,
+      arrow::MemoryPool* arrowPool);
+
+  arrow::Status write(std::shared_ptr<ColumnarBatch> cb, int64_t memLimit) override;
+
+  arrow::Status stop() override;
+
+  arrow::Status reclaimFixedSize(int64_t size, int64_t* actual) override;
+
+  arrow::Status evictRowVector(uint32_t partitionId) override;
+
+  arrow::Status evictBatch(
+      uint32_t partitionId,
+      std::ostringstream* output,
+      facebook::velox::OStreamOutputStream* out,
+      facebook::velox::RowTypePtr* rowTypePtr);
+
+ private:
+  VeloxSortBasedShuffleWriter(
+      uint32_t numPartitions,
+      std::unique_ptr<PartitionWriter> partitionWriter,
+      ShuffleWriterOptions options,
+      std::shared_ptr<facebook::velox::memory::MemoryPool> veloxPool,
+      arrow::MemoryPool* pool)
+      : VeloxShuffleWriter(numPartitions, std::move(partitionWriter), std::move(options), std::move(veloxPool), pool) {}
+
+  arrow::Status init();
+
+  arrow::Status initFromRowVector(const facebook::velox::RowVector& rv);
+
+  void setSplitState(SplitState state);
```

Comment:
`SplitState` is used to control the "spill" behavior in hash based shuffle. Perhaps we can use another one like `SortState` in sort based shuffle.

Reply:
@marin-ma I have changed ```SplitState``` to ```ShuffleState```.

Reply:
@kerwin-zk What I mean is to introduce a new state class for the sort based writer, and preserve the SplitState only for hash based writer, because the states are designed for  the concrete implementation. We'd better not to move the State class into the cpp/core module.

Reply:
OK